DependencyInjection/GuzzleHandlerStackCompilerPass.php

```diff
@@ -23,13 +24,23 @@ public function process(ContainerBuilder $container): void
     {
         foreach ($container->getDefinitions() as $clientServiceName => $definition) {
             if ($definition->getClass() === Client::class) {
-                $this->addMiddlewareToClient($container, $definition, $clientServiceName);
+                $this->processClient($container, $definition, $clientServiceName);
             } elseif ($definition->getClass() === HandlerStack::class) {
                 $this->addUniqueMethodCallsToHandlerStack($definition);
             }
         }
     }
 
+    private function processClient(ContainerBuilder $container, Definition $definition, string $clientServiceName): void
+    {
```

Comment:
No need to treat the factory case specifically, always relying on decorator should be a better option.